File: LeetCode/Medium/Add Two Numbers/Go/add_two_numbers.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func New(val int) *ListNode { return &ListNode{Val: val, Next: nil} }

func add_two_number(l1, l2 *ListNode) *ListNode {
	return _add(l1, l2, 0)
}

func _add(l1, l2 *ListNode, carry int) *ListNode {
	if l1 == nil && l2 == nil {
		return nil
	}
	if l1 == nil && l2 == nil && carry != 0 {
		return New(carry)
	}
	val := carry
	var new_node *ListNode

	if l1 != nil && l2 != nil {
		val += l1.Val + l2.Val
		l1 = l1.Next
		l2 = l2.Next
	} else if l2 == nil {
		val += l1.Val
		l1 = l1.Next
	} else {
		val += l2.Val
		l2 = l2.Next
	}
	new_node = New(val % 10)
	carry = val / 10
	new_node.Next = _add(l1, l2, carry)
	return new_node
}

func main() {
	l1 := New(2)
	l1.Next = New(4)
	l1.Next.Next = New(3)
	l2 := New(5)
	l2.Next = New(6)
	l2.Next.Next = New(4)
	fmt.Println(add_two_number(l1, l2))
}

File: LeetCode/Medium/Add Two Numbers/C/add_two_numbers.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
typedef struct ListNode ListNode;

struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2){
    ListNode* res = (ListNode*) malloc(sizeof(ListNode));
    ListNode* prev = NULL;
    ListNode* head = (ListNode*) malloc(sizeof(ListNode));
    head->val = -1;
    head->next = res;
    int carry = 0;
    while(l1 != NULL && l2 != NULL) {
        int sum = l1->val+l2->val+carry;
        carry = sum/10;
        res->val = sum%10;
        prev = res;
        res->next = (ListNode*) malloc(sizeof(ListNode));
        res = res->next;
        l1 = l1->next;
        l2 = l2->next;
    }

    while(l1 != NULL) {
        int sum = l1->val+carry;
        carry = sum/10;
        res->val = sum%10;
        prev = res;
        res->next = (ListNode*) malloc(sizeof(ListNode));
        res = res->next;
        l1 = l1->next;
    }

    while(l2 != NULL) {
        int sum = l2->val+carry;
        carry = sum/10;
        res->val = sum%10;
        prev = res;
        res->next = (ListNode*) malloc(sizeof(ListNode));
        res = res->next;
        l2 = l2->next;
    }

    while(carry > 0) {
        res->val = carry%10;
        carry = carry/10;
        prev = res;
        res->next = (ListNode*) malloc(sizeof(ListNode));
        res = res->next;
    }
    prev->next = NULL;
    free(res);
    return head->next;
}
